contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+computing recommended resources. On startup the recommender fetches
+historical resource utilization of all Pods (regardless of whether
+they use VPA) together with the history of Pod OOM events from the
+History Storage. It aggregates this data and keeps it in memory.
+
+During normal operation the recommender consumes real time updates of
+resource utilization and new events via the Master Metrics API from
+Heapster. Additionally it watches all Pods and all VPA objects in the
+cluster. For every Pod that is matched by some VPA selector the
+Recommender computes the recommended resources and sets the
+recommendation on the VPA object.
+
+It is important to realize that one VPA object has one recommendation.
+The user is expected to use one VPA to control Pods with similar
+resource usage patterns, typically a group of replicas or shards of
+a single workload.
+
+The Recommender exposes a synchronous RPC interface that takes a Pod
+Spec and the Pod metadata and returns recommended resources:
+
+```json
+GET /recommendedPodResources
+{
+  "metadata": { },
+  "spec": { }
+}
+```
+
+The response:
+
+```json
+{
+  "containers": [
+    {"name": "...", "resources": { }}
+  ]
+}
+```
+
+Notice that this API may be called for an existing Pod, as well as for a
+yet-to-be-created Pod.
+It is not guaranteed that this API will return the same recommendation for all
+Pods matched by one VPA.
+
+### Updater ###
+VPA Updater is a component responsible for applying recommended resources to
+existing Pods. In MVP (until in-place updates of Pod resources are available)
+this means evicting Pods whose recommended resources significantly diverge from
+actual resources, in order to have them recreated with the recommended
+resources.
+
+While terminating Pods is disruptive and generally undesired, it is sometimes
+justified in order to (1) avoid CPU starvation (2) reduce the risk of correlated
+OOMs across multiple Pods at random time or (3) save resources over long periods
+of time.
+
+Apart from its own policy on how often a Pod can be evicted, the Updater also
+respects the Pod disruption budget, by using Eviction API to evict pods.
+
+The Updater only touches Pods that (1) point to a VPA with updatePolicy.mode set
+to “auto” and (2) belong to a Replica Set.
+
+Note: the Updater may in future be extracted to a separate process.
+
+### Recommendation model ###
+
+VPA controls the request (memory and CPU) of containers. In MVP it always sets
+the limit to infinity. It is not yet clear whether there is a use-case for VPA
+setting the limit.
+
+The request is calculated based on analysis of the current and revious runs of
+the container and other containers with similar properties (name, image,
+command, args).
+The recommendation model (MVP) assumes that the memory and CPU consumption are
+independent random variables with distribution equal to the one observed in the
+last N days (recommended value is N=8 to capture weekly peaks).
```

Comment:
If the cluster has less than 8 days of recorded history, VPA will use whatever was collected so far.
More geneally, if the recommender has insufficent historical data to give a recommendation for a given VPA it will probably look (as we do internally) at "similar" pods running now or in the past, with a reasonable definition of "similar". Details of the algorithm tbd later, once we have agreement on the overall design.